StellarDotnetSdk.Tests/WebAuthenticationTest.cs

```diff
@@ -1691,60 +1513,186 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
+            validFor: validFor);
+        transaction.Sign(_serverKeypair);
+
+        // Grace time bounds
+        var graceValidFrom = validFrom.AddSeconds(-WebAuthentication.GracePeriod);
+
+        // Act & Assert
+        var result = WebAuthentication.ReadChallengeTransaction(
+            transaction,
+            _serverKeypair.AccountId,
+            HomeDomain,
+            WebAuthDomain,
+            // Start 1 second after the min time bounds - grace period
+            now: graceValidFrom.AddSeconds(1));
+        Assert.AreEqual(_clientKeypair.AccountId, result);
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBound()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'validFrom'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = WebAuthentication.BuildChallengeTransaction(
+            _serverKeypair,
+            _clientKeypair.AccountId,
+            HomeDomain,
+            WebAuthDomain,
+            validFrom: validFrom,
+            validFor: validFor);
+        transaction.Sign(_serverKeypair);
+
+        // Grace time bounds
+        var graceValidFor = validFor + TimeSpan.FromSeconds(WebAuthentication.GracePeriod);
+        var graceValidTill = validFrom.Add(graceValidFor);
+
+        // Act & Assert
+        Assert.ThrowsException<InvalidWebAuthenticationException>(() =>
+        {
+            WebAuthentication.ReadChallengeTransaction(
+                transaction,
+                _serverKeypair.AccountId,
+                HomeDomain,
+                WebAuthDomain,
+                // Start 1 after the max time bounds + grace period
+                now: graceValidTill.AddSeconds(1));
+        }, "Challenge transaction expired");
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBoundButWithinGracePeriod()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'validFrom'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = WebAuthentication.BuildChallengeTransaction(
+            _serverKeypair,
+            _clientKeypair.AccountId,
+            HomeDomain,
+            WebAuthDomain,
+            validFrom: validFrom,
+            validFor: validFor);
+        transaction.Sign(_serverKeypair);
+
+        // Grace time bounds
+        var graceValidFor = validFor + TimeSpan.FromSeconds(WebAuthentication.GracePeriod);
+        var graceValidTill = validFrom.Add(graceValidFor);
+
+        // Act & Assert
+        var result = WebAuthentication.ReadChallengeTransaction(
+            transaction,
+            _serverKeypair.AccountId,
+            HomeDomain,
+            WebAuthDomain,
+            // Start 1 second before the max time bounds + grace period
```

Comment:
Doesn't the max time bounds already include the grace period?

Reply:
This comment should not be there. Removed.

Reply:
Weird, I still see the comment.

Reply:
Ohh no sorry I confused this with another same comment I removed in the `_ReadChallengeTransaction` method.

To address your initial question: no, the grace period is not included in the minimum and maximum time values. In the validateTimeBounds method, the grace period is added to the calculation, but before that, the values are only what the user configured.